File: backend/apps/crm/schemas.py
from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from ninja import Schema
from pydantic import EmailStr, Field


class UserSummarySchema(Schema):
    id: int
    email: str
    first_name: str
    last_name: str


class LeadSchema(Schema):
    id: UUID
    name: str
    company: str
    email: str
    phone: str
    whatsapp: str
    country: str
    website: str
    source: str
    status: str
    priority: str
    assigned_to: UserSummarySchema | None
    lead_score: int
    estimated_value: Decimal | None
    currency: str
    notes: str
    tags: list[str]
    next_follow_up_at: datetime | None
    last_contacted_at: datetime | None
    created_at: datetime
    updated_at: datetime


class LeadCreateSchema(Schema):
    name: str = Field(min_length=1, max_length=200)
    company: str = Field(default="", max_length=200)
    email: EmailStr | None = None
    phone: str = Field(default="", max_length=40)
    whatsapp: str = Field(default="", max_length=40)
    country: str = Field(default="", max_length=100)
    website: str = ""
    source: str = "manual"
    status: str = "new"
    priority: str = "normal"
    assigned_to_id: int | None = None
    lead_score: int = Field(default=0, ge=0, le=100)
    estimated_value: Decimal | None = None
    currency: str = Field(default="LKR", min_length=3, max_length=3)
    notes: str = ""
    tags: list[str] = Field(default_factory=list)
    next_follow_up_at: datetime | None = None


class LeadUpdateSchema(LeadCreateSchema):
    pass


class LeadNoteCreateSchema(Schema):
    content: str = Field(min_length=1)
    is_pinned: bool = False


class LeadNoteSchema(Schema):
    id: UUID
    content: str
    is_pinned: bool
    created_at: datetime


class LeadTimelineSchema(Schema):
    id: UUID
    event_type: str
    description: str
    metadata: dict[str, Any]
    created_at: datetime

File: apps/dashboard/src/features/crm/schemas.ts
import {
  z,
} from "zod";

import {
  leadPriorities,
  leadSources,
  leadStatuses,
} from "./types";

const nonNegativeCount = z.number().int().nonnegative();

const userSummarySchema = z.object({
  id: z.number().int(),
  email: z.string(),
  first_name: z.string(),
  last_name: z.string(),
});

export const leadSchema = z.object({
  id: z.string().uuid(),
  name: z.string(),
  company: z.string(),
  email: z.string(),
  phone: z.string(),
  whatsapp: z.string(),
  country: z.string(),
  website: z.string(),
  source: z.enum(leadSources),
  status: z.enum(leadStatuses),
  priority: z.enum(leadPriorities),
  assigned_to: userSummarySchema.nullable(),
  lead_score: z.number().int().min(0).max(100),
  estimated_value: z.string().nullable(),
  currency: z.string(),
  notes: z.string(),
  tags: z.array(z.string()),
  next_follow_up_at: z.string().nullable(),
  last_contacted_at: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
});

export const paginatedLeadsSchema = z.object({
  items: z.array(leadSchema),
  pagination: z.object({
    page: z.number().int().positive(),
    page_size: z.number().int().positive(),
    total_items: nonNegativeCount,
    total_pages: nonNegativeCount,
  }),
});

const crmSummarySchema = z.object({
  total_leads: nonNegativeCount,
  new_leads: nonNegativeCount,
  contacted_leads: nonNegativeCount,
  follow_up_leads: nonNegativeCount,
  proposal_sent_leads: nonNegativeCount,
  negotiation_leads: nonNegativeCount,
  won_leads: nonNegativeCount,
  lost_leads: nonNegativeCount,
  spam_leads: nonNegativeCount,
  overdue_follow_ups: nonNegativeCount,
  unassigned_leads: nonNegativeCount,
  conversion_rate: z.number(),
  all_time_total_leads: nonNegativeCount,
  all_time_won_leads: nonNegativeCount,
  all_time_lost_leads: nonNegativeCount,
});

export const crmReportSchema = z.object({
  report_type: z.literal("crm"),
  environment: z.string(),
  period: z.object({
    preset: z.string(),
    date_from: z.string(),
    date_to: z.string(),
    datetime_from: z.string(),
    datetime_to: z.string(),
  }),
  generated_at: z.string(),
  timezone: z.string(),
  schema_version: z.number().int().positive(),
  data: z.object({
    summary: crmSummarySchema,
    leads_by_status: z.array(z.unknown()),
    leads_by_source: z.array(z.unknown()),
    leads_by_owner: z.array(z.unknown()),
    conversion_funnel: z.array(z.unknown()),
    monthly_lead_trend: z.array(z.unknown()),
    won_lost_trend: z.array(z.unknown()),
    estimated_value_by_currency: z.array(z.unknown()),
    metadata: z.record(
      z.string(),
      z.unknown(),
    ),
  }),
  metadata: z.object({
    foundation: z.boolean(),
    aggregation_status: z.string(),
  }),
});
